### PR TITLE
ls: stop flattening output, show the actual tree

### PR DESCRIPTION
**Pull request by iarna**
This is the infamous `debfsify` branch, finally coming home to land. It only needed a few tests tweaked that relied on specific `npm ls` output. Previously `npm` would trim and reshape the output tree to try to make it more concise. This could be confusing because it didn't represent what was actually there. Further, when modules were deduped (that is, were a shared transitive dependency) there was no way to see this was happening.

With this patch you now get output like:
```
npm@4.4.0 /Users/rebecca/code/npm-latest
├── abbrev@1.1.0
├── ansi-regex@2.1.1
├── ansicolors@0.3.2
├── ansistyles@0.1.3
├── aproba@1.1.1
├── archy@1.0.0
├── asap@2.0.5
├── chownr@1.0.1
├─┬ cmd-shim@2.0.2
│ ├── graceful-fs@4.1.11 deduped
│ └── mkdirp@0.5.1 deduped
├─┬ columnify@1.5.4
│ ├── strip-ansi@3.0.1 deduped
│ └─┬ wcwidth@1.0.0
│   └─┬ defaults@1.0.3
│     └── clone@1.0.2
```

### CODE REVIEW

File: test/tap/peer-deps-toplevel.js [hunk 1]
```diff
@@ -59,57 +72,65 @@ var expected = {
...
+    t.done()
+  })
+})
 
-        npm.commands.ls([], true, function (err, _, results) {
-          t.ifError(err, 'listed tree without problems')
+test('installs the peer dependency directory structure', function (t) {
+  common.npm(['install'], conf, function (err, code, stdout, stderr) {
+    if (err) throw err
+    t.is(code, 0, 'install ran ok even w/ missing peeer deps')
```
Comment (zkat) [inline test/tap/peer-deps-toplevel.js]: `peeeeeeeeeeeeeeeeer deps`?

File: lib/ls.js [hunk 2]
```diff
@@ -444,10 +466,11 @@ function getExtras (data) {
 }
 
 function makeParseable (data, long, dir, depth, parent, d) {
+  if (data._deduped) return []
   depth = depth || 0
   if (depth > npm.config.get('depth')) return [ makeParseable_(data, long, dir, depth, parent, d) ]
   return [ makeParseable_(data, long, dir, depth, parent, d) ]
-  .concat(Object.keys(data.dependencies || {})
+ .concat(Object.keys(data.dependencies || {})
```
Comment (zkat) [inline lib/ls.js]: isn't this indent weird?

**Review by zkat — APPROVED** [review]
LGTM 🐑

File: lib/ls.js (hunk 2 above)
Comment (legodude17) [inline lib/ls.js]: Yeah, shouldn't it be indented two spaces? Instead of outdented once.

Comment (iarna) [inline lib/ls.js]: sure looks like it, I wonder why `standard` isn't yelling…

File: test/tap/peer-deps-toplevel.js (hunk 1 above)
Comment (iarna) [inline test/tap/peer-deps-toplevel.js]: yuuuup, we still have tests for them

File: lib/ls.js (hunk 2 above)
Comment (legodude17) [inline lib/ls.js]: Maybe standard doesn't enforce spacing in method chains?

Comment (iarna) [inline lib/ls.js]: Maybe, it's awfully picky otherwise =p

Comment (legodude17) [inline lib/ls.js]: I read through the rules and there was no mention of method chains...
I don't know `standard` well enough to call whether it does or doesn't.

File: lib/ls.js [hunk 3]
```diff
@@ -354,6 +367,14 @@ function makeArchy_ (data, long, dir, depth, parent, d) {
   }
   if (data.link) out.label += ' -> ' + data.link
 
+  if (data._deduped) {
+    if (npm.color) {
+      out.label += ' ' + color.brightBlack('deduped')
+    } else {
+      out.label += ' deduped'
+    }
```
Comment (legodude17) [inline lib/ls.js]: Maybe this should be in parens? Like ` (deduped)`.

**Comment by legodude17** [comment]
Did this get merged?

**Comment by zkat** [comment]
yes

**Comment by legodude17** [comment]
🎆

**Pull request closed without merging**